Adds support to StarPU applications

weslleyspereira opened the pull request:
Related to #203.

This PR:
- [x] Adds example using StarPU.
- [x] \<T\>LAPACK plugin for StarPU.

Also:
- Fix c++14 support.
- Removes slice from GETRF level 0.

thijssteel commented:
How does StarPU fit into the bigger picture? Will we try to eventually make variants of the important routines that are optimized using StarPU? Or are we supposed to rewrite all the tiled algorithms using StarPU?

weslleyspereira commented:
New namespace `tlapack::starpu` implements:
- Matrix.
  - It has a specific way to create grids. Still not completely
functional. Currently only works if the sizes of the matrix are multiple
of the partition sizes.
  - A virtual way to create submatrices. It means: we don't change the
handle when extracting submatrices.
  - Access operators `operator()` and `operator[]`. The second only works
for 'vectors'
  - Vectors: implemented under the same matrix structure. It checks at
runtime if a type is a matrix or vector depending on the operation.
- Matrix\<T\>::data is the non-const data type of each entry. It is
different from T. `Matrix\<T\>::data` is associated to StarPU handles.
  - Operations: `{ Assign, Add, Subtract, Multiply, Divide }`. Operations
between (1) two `Matrix\<T\>::data`, and (2) one `Matrix\<T\>::data` and `T`. All
operations are done using StarPU tasks.
  - abs and sqrt are as well implemented.
  - Implicit conversion from `Matrix\<T\>::data` to T is also implemented.

sthibaul commented on include/tlapack/plugins/starpu.hpp:
```diff
@@ -0,0 +1,901 @@
+                task->cl = &cl;
+                task->handles[0] = this->handle;
+                task->synchronous = 1;
+                task->cl_arg = (void*)(&x);
```
Just to make sure: this is made synchronous because cl_arg points to x, whose liveness thus requires to wait for the task?

weslleyspereira replied:
I think I did the changes you suggested in order to be able to use asynchronous tasks. Thanks! That was one of the reasons.

However, the code still does not work with asynchronous tasks. I am not sure why.

weslleyspereira replied:
More details:
- I arrive at a segmentation fault if I put `task->synchronous = false;` on `operate_and_assign(const T& x)` or `operate_and_assign(const data& x)`.
- I am unable to identify the problem when debugging. Even when I set "STARPU_NCPU=1" and "STARPU_NCUDA=0", the debugger does not help.

sthibaul replied:
> I arrive at a segmentation fault

How does it look like exactly? (`bt full` in gdb)

weslleyspereira replied:
The lines the code stops change from execution to execution.

Using `STARPU_NCUDA=0` and `STARPU_NCPU=1` and asynchronous operation in `operate_and_assign(const T& x)`, i.e., using `task->synchronous = false` at:

https://github.com/tlapack/tlapack/blob/be72b82aca0a1aa2f3316975d3492109a137dc56/include/tlapack/plugins/starpu.hpp#L423

I noticed GDB may stop at three different places:

1. First time I obtained:

```gdb
(gdb) run
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /home/weslleyp/storage/tlapack/build_RelWithDebInfo/examples/starpu/example_starpu 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
m = 39
n = 17
method = level0

[Detaching after fork from child process 330438]
Invalid MIT-MAGIC-COOKIE-1 key[Detaching after fork from child process 330441]
[New Thread 0x7fffa0ae7000 (LWP 330442)]
[starpu][weslleyp-XPS-15-9510][initialize_lws_policy] Warning: you are running the default lws scheduler, which is not a very smart scheduler, while the system has GPUs or several memory nodes. Make sure to read the StarPU documentation about adding performance models in order to be able to use the dmda or dmdas scheduler instead.
[New Thread 0x7fff83fff000 (LWP 330443)]
[New Thread 0x7fff837fe000 (LWP 330444)]
[New Thread 0x7fff82ffd000 (LWP 330445)]
[New Thread 0x7fff827fc000 (LWP 330446)]
[New Thread 0x7fff81ffb000 (LWP 330447)]
[New Thread 0x7fff817fa000 (LWP 330448)]
[New Thread 0x7fff80ff9000 (LWP 330449)]
[New Thread 0x7fff65fff000 (LWP 330450)]
[New Thread 0x7fff657fe000 (LWP 330451)]
A = starpu::Matrix<f>( nrows = 39, ncols = 17 )

Thread 4 "CPU 0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff837fe000 (LWP 330444)]
0x00007ffff7ea2eee in _starpu_pthread_spin_lock (lock=0x30) at ./common/thread.h:86
86              if (STARPU_LIKELY(STARPU_VAL_COMPARE_AND_SWAP(&lock->taken, 0, 1) == 0))
(gdb) bt full
#0  0x00007ffff7ea2eee in _starpu_pthread_spin_lock (lock=0x30) at ./common/thread.h:86
No locals.
#1  __starpu_spin_lock (file=0x7ffff7f3c5d9 "datawizard/coherency.c", line=1287, func=<synthetic pointer>, lock=<optimized out>) at ./common/starpu_spinlock.h:98
        ret = <optimized out>
        ret = <optimized out>
        __PRETTY_FUNCTION__ = "__starpu_spin_lock"
        __ptrs = {<optimized out> <repeats 32 times>}
        __n = <optimized out>
#2  _starpu_fetch_task_input_tail (task=task@entry=0x555555ac99f0, j=j@entry=0x555555ac9c20, worker=worker@entry=0x7ffff7f9bac0 <_starpu_config+4128>) at datawizard/coherency.c:1287
        handle = <optimized out>
        node = <optimized out>
        local_replicate = 0x110
        needs_init = <optimized out>
        mode = <optimized out>
        workerid = 1
        profiling = 0
        nbuffers = <optimized out>
        descrs = <optimized out>
        index = <optimized out>
        total_size = 0
#3  0x00007ffff7f0cc6b in _starpu_cpu_driver_run_once (cpu_worker=cpu_worker@entry=0x7ffff7f9bac0 <_starpu_config+4128>) at drivers/cpu/driver_cpu.c:590
        ret = <optimized out>
        memnode = 0
        workerid = 1
        pi = {conf = 0x0, event_type = starpu_prof_tool_event_end_transfer, starpu_version = {1, 4, 99}, thread_id = -2088771584, worker_id = 1, device_number = 1, driver_type = starpu_prof_tool_driver_cpu, 
          memnode = 0, bytes_to_transfer = 1, bytes_transfered = 1, fun_ptr = 0x0}
        res = <optimized out>
        j = 0x555555ac9c20
        task = 0x0
        pending_task = 0x555555ac99f0
        rank = 0
        __func__ = "_starpu_cpu_driver_run_once"
        continuation_wake_up = <optimized out>
        __PRETTY_FUNCTION__ = "_starpu_cpu_driver_run_once"
#4  0x00007ffff7f0cf9d in _starpu_cpu_worker (arg=0x7ffff7f9bac0 <_starpu_config+4128>) at drivers/cpu/driver_cpu.c:712
        worker = 0x7ffff7f9bac0 <_starpu_config+4128>
        pi = {conf = 0x0, event_type = starpu_prof_tool_event_start_transfer, starpu_version = {1, 4, 99}, thread_id = -2088771584, worker_id = 1, device_number = 1, driver_type = starpu_prof_tool_driver_cpu, 
          memnode = 0, bytes_to_transfer = 0, bytes_transfered = 0, fun_ptr = 0x0}
#5  0x00007ffff7820609 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140735399583744, -2075732697303488179, 140737488341678, 140737488341679, 140737488341680, 140735399491840, 2075740381010454861, 2075714229153687885}, 
              mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#6  0x00007ffff7966133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.
```

2. Then I obtained:

```gdb
(gdb) run
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /home/weslleyp/storage/tlapack/build_RelWithDebInfo/examples/starpu/example_starpu 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
m = 39
n = 17
method = level0

[Detaching after fork from child process 330301]
Invalid MIT-MAGIC-COOKIE-1 key[Detaching after fork from child process 330314]
[New Thread 0x7fffa0ae7000 (LWP 330315)]
[starpu][weslleyp-XPS-15-9510][initialize_lws_policy] Warning: you are running the default lws scheduler, which is not a very smart scheduler, while the system has GPUs or several memory nodes. Make sure to read the StarPU documentation about adding performance models in order to be able to use the dmda or dmdas scheduler instead.
[New Thread 0x7fff83fff000 (LWP 330316)]
[New Thread 0x7fff837fe000 (LWP 330317)]
[New Thread 0x7fff82ffd000 (LWP 330318)]
[New Thread 0x7fff827fc000 (LWP 330319)]
[New Thread 0x7fff81ffb000 (LWP 330320)]
[New Thread 0x7fff817fa000 (LWP 330321)]
[New Thread 0x7fff80ff9000 (LWP 330322)]
[New Thread 0x7fff65fff000 (LWP 330323)]
[New Thread 0x7fff657fe000 (LWP 330324)]
A = starpu::Matrix<f>( nrows = 39, ncols = 17 )

Thread 4 "CPU 0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff837fe000 (LWP 330317)]
0x00007ffff7e30f6e in _starpu_task_uses_multiformat_handles (task=task@entry=0x555555ac99f0) at core/task.c:1631
1631                    if (_starpu_data_is_multiformat_handle(STARPU_TASK_GET_HANDLE(task, i)))
(gdb) bt full
#0  0x00007ffff7e30f6e in _starpu_task_uses_multiformat_handles (task=task@entry=0x555555ac99f0) at core/task.c:1631
        i = 1
        nbuffers = 102
#1  0x00007ffff7e70b66 in _starpu_pop_task (worker=worker@entry=0x7ffff7f9bac0 <_starpu_config+4128>) at core/sched_policy.c:1093
        task = 0x555555ac99f0
        worker_id = <optimized out>
        node = <optimized out>
        profiling = 0
        pop_start_time = {tv_sec = 140735005327392, tv_nsec = 132272}
        pick = <optimized out>
        i = <optimized out>
        nbuffers = <optimized out>
#2  0x00007ffff7e9cba1 in _starpu_get_worker_task (worker=worker@entry=0x7ffff7f9bac0 <_starpu_config+4128>, workerid=workerid@entry=1, memnode=memnode@entry=0) at drivers/driver_common/driver_common.c:427
        task = <optimized out>
        keep_awake = <optimized out>
        __func__ = "_starpu_get_worker_task"
#3  0x00007ffff7f0ca4c in _starpu_cpu_driver_run_once (cpu_worker=cpu_worker@entry=0x7ffff7f9bac0 <_starpu_config+4128>) at drivers/cpu/driver_cpu.c:606
        memnode = 0
        workerid = 1
        pi = {conf = 0x0, event_type = starpu_prof_tool_event_driver_init_end, starpu_version = {1, 4, 99}, thread_id = -2088771584, worker_id = 1, device_number = 0, driver_type = starpu_prof_tool_driver_cpu, 
          memnode = 4294967295, bytes_to_transfer = 0, bytes_transfered = 0, fun_ptr = 0x0}
        res = <optimized out>
        j = <optimized out>
        task = 0x0
        pending_task = 0x0
        rank = 0
        __func__ = "_starpu_cpu_driver_run_once"
        continuation_wake_up = <optimized out>
        __PRETTY_FUNCTION__ = "_starpu_cpu_driver_run_once"
#4  0x00007ffff7f0cf9d in _starpu_cpu_worker (arg=0x7ffff7f9bac0 <_starpu_config+4128>) at drivers/cpu/driver_cpu.c:712
        worker = 0x7ffff7f9bac0 <_starpu_config+4128>
        pi = {conf = 0x0, event_type = starpu_prof_tool_event_start_transfer, starpu_version = {1, 4, 99}, thread_id = -2088771584, worker_id = 1, device_number = 1, driver_type = starpu_prof_tool_driver_cpu, 
          memnode = 0, bytes_to_transfer = 0, bytes_transfered = 0, fun_ptr = 0x0}
#5  0x00007ffff7820609 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140735399583744, -4789584980818873970, 140737488341678, 140737488341679, 140737488341680, 140735399491840, 4789858763391545742, 4789603295442022798}, 
              mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#6  0x00007ffff7966133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.
```

3. Then:

```gdb
(gdb) run
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /home/weslleyp/storage/tlapack/build_RelWithDebInfo/examples/starpu/example_starpu 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
m = 39
n = 17
method = level0

[Detaching after fork from child process 330633]
Invalid MIT-MAGIC-COOKIE-1 key[Detaching after fork from child process 330636]
[New Thread 0x7fffa0ae7000 (LWP 330637)]
[starpu][weslleyp-XPS-15-9510][initialize_lws_policy] Warning: you are running the default lws scheduler, which is not a very smart scheduler, while the system has GPUs or several memory nodes. Make sure to read the StarPU documentation about adding performance models in order to be able to use the dmda or dmdas scheduler instead.
[New Thread 0x7fff83fff000 (LWP 330638)]
[New Thread 0x7fff837fe000 (LWP 330639)]
[New Thread 0x7fff82ffd000 (LWP 330640)]
[New Thread 0x7fff827fc000 (LWP 330641)]
[New Thread 0x7fff81ffb000 (LWP 330642)]
[New Thread 0x7fff817fa000 (LWP 330643)]
[New Thread 0x7fff80ff9000 (LWP 330644)]
[New Thread 0x7fff65fff000 (LWP 330645)]
[New Thread 0x7fff657fe000 (LWP 330646)]
A = starpu::Matrix<f>( nrows = 39, ncols = 17 )

Thread 4 "CPU 0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff837fe000 (LWP 330639)]
_starpu_fetch_data_on_node (handle=handle@entry=0x555555ac7dd0, node=node@entry=32767, dst_replicate=<optimized out>, mode=mode@entry=STARPU_W, detached=detached@entry=0, task=task@entry=0x555555ac99f0, 
    is_prefetch=STARPU_FETCH, async=1, callback_func=0x7ffff7ea0270 <_starpu_fetch_task_input_cb>, callback_arg=0x7ffff7f9bac0 <_starpu_config+4128>, prio=0, origin=0x7ffff7f3c702 "_starpu_fetch_task_input")
    at datawizard/coherency.c:867
867                             dst_replicate->refcnt++;
(gdb) bt full
#0  _starpu_fetch_data_on_node (handle=handle@entry=0x555555ac7dd0, node=node@entry=32767, dst_replicate=<optimized out>, mode=mode@entry=STARPU_W, detached=detached@entry=0, task=task@entry=0x555555ac99f0, 
    is_prefetch=STARPU_FETCH, async=1, callback_func=0x7ffff7ea0270 <_starpu_fetch_task_input_cb>, callback_arg=0x7ffff7f9bac0 <_starpu_config+4128>, prio=0, origin=0x7ffff7f3c702 "_starpu_fetch_task_input")
    at datawizard/coherency.c:867
        __func__ = "_starpu_fetch_data_on_node"
        __PRETTY_FUNCTION__ = "_starpu_fetch_data_on_node"
        r = <optimized out>
        ret = <optimized out>
#1  0x00007ffff7ea3215 in _starpu_fetch_task_input (task=0x555555ac99f0, j=0x555555ac9c20, async=async@entry=1) at datawizard/coherency.c:1195
        ret = <optimized out>
        handle = 0x555555ac7dd0
        node = 32767
        local_replicate = <optimized out>
        mode = STARPU_W
        worker = 0x7ffff7f9bac0 <_starpu_config+4128>
        workerid = 1
        profiling = <optimized out>
        descrs = 0x555555ac9c90
        nbuffers = 4294955272
        nacquires = 0
        index = 0
        __func__ = "_starpu_fetch_task_input"
        __PRETTY_FUNCTION__ = "_starpu_fetch_task_input"
        index2 = <optimized out>
#2  0x00007ffff7f0cb36 in _starpu_cpu_driver_run_once (cpu_worker=cpu_worker@entry=0x7ffff7f9bac0 <_starpu_config+4128>) at drivers/cpu/driver_cpu.c:677
        memnode = 0
        workerid = 1
        pi = {conf = 0x0, event_type = starpu_prof_tool_event_end_transfer, starpu_version = {1, 4, 99}, thread_id = -2088771584, worker_id = 1, device_number = 1, driver_type = starpu_prof_tool_driver_cpu, 
          memnode = 0, bytes_to_transfer = 0, bytes_transfered = 0, fun_ptr = 0x0}
        res = <optimized out>
        j = <optimized out>
        task = <optimized out>
        pending_task = <optimized out>
        rank = <optimized out>
        __func__ = "_starpu_cpu_driver_run_once"
        continuation_wake_up = <optimized out>
        __PRETTY_FUNCTION__ = "_starpu_cpu_driver_run_once"
#3  0x00007ffff7f0cf9d in _starpu_cpu_worker (arg=0x7ffff7f9bac0 <_starpu_config+4128>) at drivers/cpu/driver_cpu.c:712
        worker = 0x7ffff7f9bac0 <_starpu_config+4128>
        pi = {conf = 0x0, event_type = starpu_prof_tool_event_start_transfer, starpu_version = {1, 4, 99}, thread_id = -2088771584, worker_id = 1, device_number = 1, driver_type = starpu_prof_tool_driver_cpu, 
          memnode = 0, bytes_to_transfer = 0, bytes_transfered = 0, fun_ptr = 0x0}
#4  0x00007ffff7820609 in start_thread (arg=<optimized out>) at pthread_create.c:477
        ret = <optimized out>
        pd = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140735399583744, 7161945090767168267, 140737488341678, 140737488341679, 140737488341680, 140735399491840, -7162040752430144757, -7161928363691105525}, 
              mask_was_saved = 0}}, priv = {pad = {0x0, 0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = 0
#5  0x00007ffff7966133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
No locals.
```

weslleyspereira replied:
I noticed it can stop at other positions as well. So maybe there is nothing special with those 3 locals above unfortunately.

sthibaul replied:
Wondering: what is the liveness of the cl structure? It has to survive at least until the callback of the task.

weslleyspereira replied:
It may be it then, thanks!

I am working on a design with static codelets. May I consider that a codelet will be never modified inside a task?

If they are modified, then I need to allocate them dynamically as well.

weslleyspereira replied:
Fixed in bd201e3

sthibaul replied:
StarPU may normalize some fields in the codelet when submitting a task, so making it const may break. But otherwise you can use the codelet structure for various tasks, that will be fine, so you can make it static.

weslleyspereira replied:
Is there a way to normalize those fields before submitting any task, so that the result is truly static? Thanks

weslleyspereira replied:
It seems `_starpu_driver_start_job()` also needs to modify the codelet. I see the line `cl->per_worker_stats[workerid]++;`.

weslleyspereira replied:
I noticed that the field `per_worker_stats` is only used for profiling. The documentation also says "Optional array for statistics".  So, I assume this can be made optional in StarPU. I would like to propose a flag to disable it, something like `STARPU_UPDATE_WORKER_STATS` or another name that makes sense. With this change, codelets could be implemented as static constant expressions. Right now, as I noticed in the StarPU examples and Chameleon, codelets needs to be "global" non-const variables. By "global", I mean: "not associated to the local scope where the task is defined". 

Global variables are bad for a template library such as \<T\>LAPACK. The reason: global variables cannot be on header files, they should be either on the application executable or inside the compiled library. In the current state of StarPU, I need to dynamically allocate space for codelets.
- First, they are not that small (704B).
- Second, I don't feel good on using dynamic allocation on variables that are almost constants as the case of the codlets.

What do you think @sthibaul? I can write a PR in the StarPU repo.

weslleyspereira replied:
Great, thanks! I will look into it and proposed a PR.

sthibaul commented on include/tlapack/plugins/starpu.hpp:
```diff
@@ -0,0 +1,927 @@
+
+                // Allocate space for x
+                void* x_copy;
+                const int info = starpu_malloc(&x_copy, sizeof(T));
```
you do not need to starpu_malloc (i.e. pin) it: since it will be passed as is, without any cuda transfer or something, you can just malloc it.

weslleyspereira commented:
I made some performance measurements using the script
https://github.com/tlapack/tlapack/blob/29c3a152d2ccb33cde5edc3e3b1d111ed127c35f/examples/starpu/run_potrf.ipynb. I want to share some of the results that I found.

The machine:
- Ubuntu 20.04.6 LTS x86_64
- CPU: 11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz (8 cores)
- GPU: NVIDIA GeForce RTX 3050 (2048 cores, 4GB). Driver version: 530.30.02

Packages and configurations:
- CUDA 11.0 (cuBLAS 11.2.0 cuSOLVER 10.6.0).
- MKL 2023.1.0.
- GNU compilers, version 9.4.0.
- StarPU uses the scheduler: `dmdas`.

Main codes:
1. MKL potrf in https://github.com/tlapack/tlapack/blob/29c3a152d2ccb33cde5edc3e3b1d111ed127c35f/examples/starpu/compare_with_mkl.cpp
2. StarPU example in https://github.com/starpu-runtime/starpu/blob/master/examples/cholesky/cholesky_implicit.c
3. \<T\>LAPACK example in https://github.com/tlapack/tlapack/blob/29c3a152d2ccb33cde5edc3e3b1d111ed127c35f/examples/starpu/example_potrf.cpp

Problem:
- n-by-n matrix with `n = 7680`.
- single precision.

### Test 1: StarPU and MKL. No GPUs.

\<T\>LAPACK example with various block sizes:
![cpuOnlyGFLOPS](https://user-images.githubusercontent.com/16517625/236561290-31e685cf-c29d-4054-b3b0-f721d665d3b8.png)
Block size 512 is the best choice.

#### Results

1. MKL potrf:
    Best time =  0.256 s.
    GFLOPS =  589.0.

2. StarPU example with block size 512:
    Best time =  0.291 s.
    GFLOPS =  519.3.

3. \<T\>LAPACK example with block size 512:
    Best time =  0.316 s.
    GFLOPS =  477.9.

### Test 2: StarPU with MKL, cuBLAS and cuSOLVER.

\<T\>LAPACK example with various block sizes:
[GFLOPS.pdf](https://github.com/tlapack/tlapack/files/11409919/GFLOPS.pdf)
Block size 512 is the best choice.

#### Results

1. StarPU example with block size 512:
    Best time =  0.082 s.
    GFLOPS =  1835.1.

2. \<T\>LAPACK example with block size 512:
    Best time =  0.104 s.
    GFLOPS =  1445.6.

sthibaul commented:
> I made some performance measurements using the script https://github.com/tlapack/tlapack/blob/29c3a152d2ccb33cde5edc3e3b1d111ed127c35f/examples/starpu/run_potrf.ipynb. I want to share some of the results that I found.

So this is a mere dense tiled Cholesky factorization? In principle the <T>LAPACK performance should be really close to the starpu-example performance, I'm wondering what is different. Did you set priorities in the <T>LAPACK algorithm? They are really useful to drive the critical path and unlock parallelism. You can try to pass -no-prio to the starpu-example to disable them there to see the difference.

Which starpu scheduler did you use? For the GPU case it's really useful to try with `STARPU_SCHED=dmdas`

weslleyspereira commented:
> So this is a mere dense tiled Cholesky factorization? In principle the LAPACK performance should be really close to the starpu-example performance, I'm wondering what is different. Did you set priorities in the LAPACK algorithm? They are really useful to drive the critical path and unlock parallelism. You can try to pass -no-prio to the starpu-example to disable them there to see the difference.

Thanks for the comments! I tried to set `-no-prio` to the StarPU example. I couldn't find any changes in performance. 

> Which starpu scheduler did you use? For the GPU case it's really useful to try with `STARPU_SCHED=dmdas`

Yes, I am already using `dmdas` as the scheduler.

weslleyspereira commented:
I am ready to merge this PR because it is already too big and the tests have satisfactory results.

Next steps could be:
1. Run a profiler to try to identify if we could improve the integration with StarPU.
2. Work more on `examples/starpu/example_multishiftqr.cpp`. In its current state, we still perform many entry A(i,j) accesses.
3. Add more examples with StarPU.
4. Test the StarPU Matrix class `starpu::Matrix` in our test suite.
